src/function/mtk.jl

```diff
@@ -7,101 +7,37 @@ AutoModelingToolkit() = AutoModelingToolkit(false, false)
 
-    hess_prototype, cons_jac_prototype, cons_hess_prototype = nothing, nothing, nothing
+    sys = ModelingToolkit.modelingtoolkitize(OptimizationProblem(f, x, p;
+                                                                 lcons = fill(0.0,
+                                                                              num_cons),
+                                                                 ucons = fill(0.0,
+                                                                              num_cons)))
+    sys = ModelingToolkit.structural_simplify(sys)
+    f = OptimizationProblem(sys, x, p, grad = true, hess = true,
+                            sparse = adtype.obj_sparse, cons_j = true, cons_h = true,
+                            cons_sparse = adtype.cons_sparse).f
 
-    if f.grad === nothing
-        grad_oop, grad_iip = ModelingToolkit.generate_gradient(sys, expression = Val{false})
-        grad(J, u) = (grad_iip(J, u, p); J)
-    else
-        grad = (G, θ, args...) -> f.grad(G, θ, p, args...)
-    end
-
-    if f.hess === nothing
-        hess_oop, hess_iip = ModelingToolkit.generate_hessian(sys, expression = Val{false},
-                                                              sparse = adtype.obj_sparse)
-        hess(H, u) = (hess_iip(H, u, p); H)
-    else
-        hess = (H, θ, args...) -> f.hess(H, θ, p, args...)
-    end
-
-    if f.hv === nothing
-        hv = function (H, θ, v, args...)
-            res = adtype.obj_sparse ? hess_prototype : ArrayInterfaceCore.zeromatrix(θ)
-            hess(res, θ, args...)
-            H .= res * v
-        end
-    else
-        hv = f.hv
-    end
-
-    expr = symbolify(ModelingToolkit.Symbolics.toexpr(ModelingToolkit.equations(sys)))
-    pairs_arr = p isa SciMLBase.NullParameters ?
-                [Symbol(_s) => Expr(:ref, :x, i) for (i, _s) in enumerate(sys.states)] :
-                [
-        [Symbol(_s) => Expr(:ref, :x, i) for (i, _s) in enumerate(sys.states)]...,
-        [Symbol(_p) => p[i] for (i, _p) in enumerate(sys.ps)]...,
-    ]
-    rep_pars_vals!(expr, pairs_arr)
+    grad = (G, θ, args...) -> f.grad(G, θ, p, args...)
 
-    if f.cons === nothing
-        cons = nothing
-        cons_exprs = nothing
-    else
-        cons = (res, θ) -> f.cons(res, θ, p)
-        cons_oop = (x, p) -> (_res = zeros(eltype(x), num_cons); f.cons(_res, x, p); _res)
+    hess = (H, θ, args...) -> f.hess(H, θ, p, args...)
 
-        cons_sys = ModelingToolkit.modelingtoolkitize(NonlinearProblem(cons_oop, x, p))
-        cons_eqs = ModelingToolkit.equations(cons_sys)
-        cons_exprs = map(cons_eqs) do cons_eq
-            e = symbolify(ModelingToolkit.Symbolics.toexpr(cons_eq))
-            rep_pars_vals!(e, pairs_arr)
-            return Expr(:call, :(==), e.args[3], :0)
-        end
+    hv = function (H, θ, v, args...)
+        res = adtype.obj_sparse ? hess_prototype : ArrayInterfaceCore.zeromatrix(θ)
+        hess(res, θ, args...)
+        H .= res * v
```

Comment:
why this? Why is it not just generating an hv in the modelingtoolkitize? This is a very inefficient one.